### PR TITLE
e2e endpoint slices with duplicate addresses

### PR DESCRIPTION
**Pull request by aojea**
/kind bug

As documented for EndpointSlices in https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#duplicate-endpoints

> Due to the nature of EndpointSlice changes, endpoints may be represented in more than one EndpointSlice at the same time. This naturally occurs as changes to different EndpointSlice objects can arrive at the Kubernetes client watch / cache at different times.

Dataplane implementations should aggregate all the endpoint slices related to the service and build the set of endpoints, this e2e creates two endpoint slices with one different endpoint in each and verifies that the dataplane forwards both endpoints correctly

```release-note
NONE
```

### CODE REVIEW

**Comment by aojea** [comment]
This is a consequence of https://github.com/kubernetes/kubernetes/pull/114155, it seems that some implementations doesn't support multiple EndpointSlices and only use one of them

File: test/e2e/network/endpointslice.go [hunk 1]
```diff
@@ -524,6 +526,209 @@ var _ = common.SIGDescribe("EndpointSlice", func() {
 		framework.ExpectNoError(err)
 		framework.ExpectEqual(len(epsList.Items), 0, "filtered list should have 0 items")
 	})
+
+	ginkgo.It("should forward traffic to a Service with two custom EndpointSlices with duplicate addresses", func() {
```
Comment (danwinship) [inline test/e2e/network/endpointslice.go:530]: This doesn't describe the test well (or differentiate it from the other one). Maybe
`"should support a Service with multiple ports specified in multiple EndpointSlices"`?

The fact that the slices are "custom" is not really relevant. In theory the EndpointSlice controller could have created them that way.

Also, FTR, I'm not certain that this behavior is actually expected. I would have said that it was erroneous to have multiple EndpointSlices for the same service with different Ports. (Though, I guess, what's the point of having the Ports in the EndpointSlice if they can't be different from the Ports in the Service?)

File: test/e2e/network/endpointslice.go [hunk 2]
```diff
@@ -524,6 +526,209 @@ var _ = common.SIGDescribe("EndpointSlice", func() {
...
+		e2epod.NewPodClient(f).CreateSync(pausePod0)
+
+		dest1 := net.JoinHostPort(svc.Spec.ClusterIP, "80")
+		dest2 := net.JoinHostPort(svc.Spec.ClusterIP, "81")
+		execHostnameTest(*pausePod0, dest1, serverPod.Name)
+		execHostnameTest(*pausePod0, dest2, serverPod.Name)
+
+	})
+
+	ginkgo.It("should forward traffic to a Service with custom endpoints in different EndpointSlices", func() {
```
Comment (danwinship) [inline test/e2e/network/endpointslice.go:629]: `"should support a Service with multiple endpoint IPs specified in multiple EndpointSlices"`

File: test/e2e/network/endpointslice.go [hunk 3]
```diff
@@ -524,6 +526,209 @@ var _ = common.SIGDescribe("EndpointSlice", func() {
...
+			e2epod.NewAgnhostContainer("container-handle-9090-request", nil, port9090, "netexec", "--http-port", "9090", "--udp-port", "-1"),
+		)
+
+		pod := e2epod.NewPodClient(f).CreateSync(serverPod)
+
+		if pod.Status.PodIP == "" {
+			framework.Failf("PodIP not assigned for pod %s", pod.Name)
+		}
+
+		addresType := discoveryv1.AddressTypeIPv4
```
Comment (danwinship) [inline test/e2e/network/endpointslice.go:578]:
```suggestion
		addressType := discoveryv1.AddressTypeIPv4
```

File: test/e2e/network/endpointslice.go (hunk 1 above)
Comment (aojea) [reply test/e2e/network/endpointslice.go:530]: this is a consequence of this https://github.com/kubernetes/kubernetes/pull/114155 , basically, if you create an Endpoint with multiple subsets, the mirroring controller creates this weird Slices
I thought it was a simpler problem, at this point I should open an issue and explain it correctly

File: test/e2e/network/endpointslice.go (hunk 2 above)
Comment (aojea) [reply test/e2e/network/endpointslice.go:629]: done

File: test/e2e/network/endpointslice.go (hunk 1 above)
Comment (aojea) [reply test/e2e/network/endpointslice.go:530]: done

**Comment by aojea** [comment]
> fatal: unable to access 'https://github.com/kubernetes/kubernetes.git/': Could not resolve host: github.com

/retest

**Comment by aojea** [comment]
/retest

ping @danwinship @robscott 

can I get LGTM?

File: test/e2e/network/endpointslice.go [hunk 4]
```diff
@@ -524,6 +526,209 @@ var _ = common.SIGDescribe("EndpointSlice", func() {
...
+		eps2 := epsTemplate.DeepCopy()
+		eps2.Ports = []discoveryv1.EndpointPort{{
+			Name:     pointer.String("port81"),
+			Port:     pointer.Int32(9090),
+			Protocol: &tcpProtocol,
+		}}
+
+		_, err = f.ClientSet.DiscoveryV1().EndpointSlices(ns).Create(context.TODO(), eps2, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
```
Comment (squeed) [inline test/e2e/network/endpointslice.go:616]: Do you want to allow time for endpoints to propagate?

File: test/e2e/network/endpointslice.go [hunk 5]
```diff
@@ -524,6 +526,209 @@ var _ = common.SIGDescribe("EndpointSlice", func() {
...
+		// connect to the service must work
+		ginkgo.By("Creating a pause pods that will try to connect to the webserver")
+		pausePod0 := e2epod.NewAgnhostPod(ns, "pause-pod-0", nil, nil, nil)
+		e2epod.NewPodClient(f).CreateSync(pausePod0)
+
+		dest1 := net.JoinHostPort(svc.Spec.ClusterIP, "80")
+		dest2 := net.JoinHostPort(svc.Spec.ClusterIP, "81")
+		execHostnameTest(*pausePod0, dest1, serverPod.Name)
+		execHostnameTest(*pausePod0, dest2, serverPod.Name)
```
Comment (squeed) [inline test/e2e/network/endpointslice.go:625]: In the off-chance that this is non-deterministic, should we perform more than one connectivity test?

Comment (aojea) [reply test/e2e/network/endpointslice.go:625]: the pods are created with `CreateSync` that blocks until the pods are ready.
There is only one service so it has to be processed at the same time by whatever implements the Service -> Pod translation, if it delays more than 2 minutes (timeout in execHostnameTest) to configure 2 ports in the same Service I would say that is legit to fail ;)

**Comment by aojea** [comment]
```


k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler: 
```
/test pull-kubernetes-unit

**Pull request merged**